src/main/scala/beam/agentsim/agents/PersonAgent.scala

```diff
@@ -1452,14 +1468,33 @@ class PersonAgent(
     )
     eventsManager.processEvent(odSkimmerEvent)
     if (beamServices.beamConfig.beam.exchange.output.activitySimSkimsEnabled) {
-      val (origin, destination) = beamScenario.exchangeGeoMap match {
-        case Some(geoMap) =>
-          val origGeo = geoMap.getTAZ(origCoord)
-          val destGeo = geoMap.getTAZ(destCoord)
-          (origGeo.tazId.toString, destGeo.tazId.toString)
-        case None =>
-          (odSkimmerEvent.origin, odSkimmerEvent.destination)
-      }
+      val startLink = currentActivity.getLinkId
+      val endLinkOption = nextActivity.map(_.getLinkId)
+      val (origin, destination) =
+        if (beamScenario.tazTreeMap.tazListContainsGeoms && endLinkOption.isDefined) {
+          val origGeo = beamScenario.tazTreeMap
+            .getTAZfromLink(startLink)
+            .map(_.tazId.toString)
+            .getOrElse("NA")
+          val destGeo = beamScenario.tazTreeMap
+            .getTAZfromLink(endLinkOption.get)
+            .map(_.tazId.toString)
+            .getOrElse("NA")
+          (origGeo, destGeo)
+        } else {
+
+          beamScenario.exchangeGeoMap match {
+            case Some(geoMap) =>
+              val origGeo = geoMap.getTAZ(origCoord)
+              val destGeo = geoMap.getTAZ(destCoord)
+              logger.warn(
+                s"Could not find linkId->taz mapping so falling back on default: Trip from $origGeo to $destGeo"
+              )
```

Comment:
How many of the following log entries there will be?
Maybe it is worth writing these coordinate pairs into a separate file and putting into beamLog just a warning about a number of these pairs?
Otherwise beamLog will be too big to work with it without any benefits for us.